src/account/operations/transaction/high_level/create_alias.rs

```diff
@@ -0,0 +1,169 @@
+    },
+};
+use serde::{Deserialize, Serialize};
+
+use crate::{
+    account::{handle::AccountHandle, types::Transaction, OutputData, TransactionOptions},
+    Error,
+};
+
+/// Address and alias for `create_alias_output()`
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AliasOutputOptions {
+    /// Bech32 encoded address which will control the alias. Default will use the first
+    /// address of the account
+    pub address: Option<String>,
+    /// Immutable alias metadata
+    #[serde(rename = "immutableMetadata")]
+    pub immutable_metadata: Option<Vec<u8>>,
+    /// Alias metadata
+    pub metadata: Option<Vec<u8>>,
+}
+
+/// Dto for aliasOptions
+#[derive(Debug, Clone, Serialize, Deserialize)]
+pub struct AliasOutputOptionsDto {
+    /// Bech32 encoded address which will control the alias. Default will use the first
+    /// address of the account
+    pub address: Option<String>,
+    /// Immutable alias metadata, hex encoded bytes
+    #[serde(rename = "immutableMetadata")]
+    pub immutable_metadata: Option<String>,
+    /// Alias metadata, hex encoded bytes
+    pub metadata: Option<String>,
+}
+
+impl TryFrom<&AliasOutputOptionsDto> for AliasOutputOptions {
+    type Error = crate::Error;
+
+    fn try_from(value: &AliasOutputOptionsDto) -> crate::Result<Self> {
+        Ok(Self {
+            address: value.address.clone(),
+            immutable_metadata: match &value.immutable_metadata {
+                Some(metadata) => {
+                    Some(prefix_hex::decode(metadata).map_err(|_| DtoError::InvalidField("immutable_metadata"))?)
+                }
+                None => None,
+            },
+            metadata: match &value.metadata {
+                Some(metadata) => Some(prefix_hex::decode(metadata).map_err(|_| DtoError::InvalidField("metadata"))?),
+                None => None,
+            },
+        })
+    }
+}
+
+impl AccountHandle {
+    /// Function to create an alias output.
+    /// ```ignore
+    /// let alias_options = AliasOutputOptions {
+    ///     address: None,
+    ///     immutable_metadata: Some(b"some immutable alias metadata".to_vec()),
+    ///     metadata: Some(b"some alias metadata".to_vec()),
+    /// };
+    ///
+    /// let transaction = account.create_alias_output(alias_options, None).await?;
+    /// println!(
+    ///     "Transaction: {} Block sent: http://localhost:14265/api/core/v2/blocks/{}",
+    ///     transaction.transaction_id,
+    ///     transaction.block_id.expect("no block created yet")
+    /// );
+    /// ```
+    pub async fn create_alias_output(
+        &self,
+        alias_output_options: Option<AliasOutputOptions>,
+        options: Option<TransactionOptions>,
+    ) -> crate::Result<Transaction> {
+        let prepared_transaction = self.prepare_create_alias_output(alias_output_options, options).await?;
+        self.sign_and_submit_transaction(prepared_transaction).await
+    }
+
+    pub(crate) async fn prepare_create_alias_output(
```

Comment:
You can reorganise this method by having only one `if let Some(alias_output_options) = alias_output_options.take() {` so you can remove all the `and_then`, `as_ref` and `clone`

Reply:
How? I changed the lower part, but we need the controller_address earlier, just the metadata is optional

Reply:
You don't really need it earlier, it's a builder add unlock method, just create the basic builder first and deal with the rest in the if

Reply:
But then we have ```Rust
.add_unlock_condition(UnlockCondition::StateControllerAddress(
                    StateControllerAddressUnlockCondition::new(controller_address),
                ))
                .add_unlock_condition(UnlockCondition::GovernorAddress(GovernorAddressUnlockCondition::new(
                    controller_address,
                )))
```
twice, just with maybe another address?

Reply:
Something like this
```rust
pub(crate) async fn prepare_create_alias_output(
        &self,
        alias_output_options: Option<AliasOutputOptions>,
        options: Option<TransactionOptions>,
    ) -> crate::Result<PreparedTransactionData> {
        log::debug!("[TRANSACTION] prepare_create_alias_output");
        let rent_structure = self.client.get_rent_structure().await?;

        let mut alias_output_builder =
            AliasOutputBuilder::new_with_minimum_storage_deposit(rent_structure, AliasId::null())?
                .with_state_index(0)
                .with_foundry_counter(0);

        if let Some(alias_output_options) = alias_output_options {
            let controller_address = match alias_output_options.address {
                Some(bech32_address) => Address::try_from_bech32(bech32_address)?.1,
                None => {
                    self.public_addresses()
                        .await
                        .first()
                        .ok_or(Error::FailedToGetRemainder)?
                        .address
                        .inner
                }
            };

            alias_output_builder = alias_output_builder
                .add_unlock_condition(UnlockCondition::StateControllerAddress(
                    StateControllerAddressUnlockCondition::new(controller_address),
                ))
                .add_unlock_condition(UnlockCondition::GovernorAddress(GovernorAddressUnlockCondition::new(
                    controller_address,
                )));

            if let Some(immutable_metadata) = alias_output_options.immutable_metadata {
                alias_output_builder = alias_output_builder
                    .add_immutable_feature(Feature::Metadata(MetadataFeature::new(immutable_metadata)?));
            }

            if let Some(metadata) = alias_output_options.metadata {
                alias_output_builder =
                    alias_output_builder.add_feature(Feature::Metadata(MetadataFeature::new(metadata)?));
            }

            if let Some(state_metadata) = alias_output_options.state_metadata {
                alias_output_builder = alias_output_builder.with_state_metadata(state_metadata);
            }
        }
        

        let outputs = vec![alias_output_builder.finish_output()?];

        self.prepare_transaction(outputs, options).await
    }
```

Reply:
If alias_output_options is none it will fail, because no StateControllerAddress or GovernorAddress was added

Reply:
Right